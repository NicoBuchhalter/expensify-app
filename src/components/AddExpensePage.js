import React from 'react';
import ExpenseForm from './ExpenseForm';
import LoadingPage from './LoadingPage';
import {connect} from 'react-redux';
import { startAddExpense } from '../actions/expenses';

export class AddExpensePage extends React.Component {

  state = {
    loading: false
  }

  onSubmit = (expense) => {
    this.setState({ loading: true });
    this.props.startAddExpense(expense).then(() => {
      this.props.history.push('/');
    })
  };

  render() {
    return(
      <div>
        { this.state.loading ? 
          (
            <LoadingPage/>
          ) :
          (
            <div>
              <div className='page-header'>
                <div className='content-container'>
                  <h1 className='page-header__title'>Add Expense</h1>
                </div>
              </div>
              <div className='content-container'>
                <ExpenseForm 
                  onSubmit={this.onSubmit}
                  submitText='Save Expense'
                />
              </div>
            </div>
          )
        }
      </div>
    )
  }

}


const mapDispatchToProps = (dispatch) => ({
  startAddExpense: (expense) => dispatch(startAddExpense(expense))
});


export default connect(undefined, mapDispatchToProps)(AddExpensePage);