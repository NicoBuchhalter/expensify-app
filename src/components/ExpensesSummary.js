import React from 'react';
import { connect } from 'react-redux';
import selectExpensesTotal from '../selectors/expenses-total';
import selectExpenses from '../selectors/expenses';
import numeral from 'numeral';

export const ExpensesSummary = ({ expenseCount, expensesTotal}) => {
  const expenseWord = expenseCount === 1 ? 'expense' : 'expenses';
  const formattedTotal = numeral(expensesTotal / 100).format('$0,0.00');
  return (
    <div>
      <h1>Viewing {expenseCount} {expenseWord} totalling {formattedTotal}</h1>
    </div>
  );
};

const mapStateToProps = (state) => {
  const expenses = selectExpenses(state.expenses, state.filters)
  return {
    expenseCount: expenses.length,
    expensesTotal: selectExpensesTotal(expenses)
  };
};

export default connect(mapStateToProps)(ExpensesSummary);
