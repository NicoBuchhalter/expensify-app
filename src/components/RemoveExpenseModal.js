import React from 'react';
import Modal from 'react-modal';


const RemoveExpenseModal = (props) => (
  <Modal 
    isOpen={props.isModalOpen} 
    contentLabel="Remove Expense"
    closeTimeoutMS={200}
    className="modal"
  >
    <h3 className='modal__title'>{props.expenseDescription}</h3>
    <p className='modal__body'>Are you sure you want to remove this expense?</p>
    <div className='modal__buttons'>
      <button className='button button--secondary' onClick={props.handleRemoveExpense}>Remove</button>
      <button className='button' onClick={props.handleCloseModal}>Cancel</button>
    </div>
  </Modal>
)

export default RemoveExpenseModal;