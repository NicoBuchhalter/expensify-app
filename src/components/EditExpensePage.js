import React from 'react';
import {connect} from 'react-redux';
import RemoveExpenseModal from './RemoveExpenseModal';
import ExpenseForm from './ExpenseForm';
import LoadingPage from './LoadingPage';
import { startEditExpense, startRemoveExpense } from '../actions/expenses';

export class EditExpensePage extends React.Component {

  state = {
    isModalOpen: false,
    loading: false
  }

  onSubmit = (expense) => {
    this.setState({ loading: true });
    this.props.startEditExpense(this.props.expense.id, expense).then(() => {
      this.props.history.push('/');
    })
  };

  onRemove = () => {
    this.setState({ isModalOpen: false, loading: true });
    this.props.startRemoveExpense(this.props.expense.id).then(() => {
      this.props.history.push('/');
    })
  };

  handleOpenModal = () => {
    this.setState({ isModalOpen: true });
  }

  closeModal = () => {
    this.setState({ isModalOpen: false })
  }

  render() {
    return (
      <div>
        {!this.state.loading ? (
          <div>
            <div className='page-header'>
              <div className='content-container'>
                <h1 className='page-header__title'>Edit Expense</h1>
              </div>
            </div>
            <div className='content-container'>
              <ExpenseForm
                expense={this.props.expense}
                onSubmit={this.onSubmit}
                submitText='Save Expense'
              />
              <button className='button--secondary' onClick={this.handleOpenModal}>Remove Expense</button>
            </div>
            <RemoveExpenseModal 
              isModalOpen={this.state.isModalOpen} 
              expenseDescription={this.props.expense.description}
              handleRemoveExpense={this.onRemove}
              handleCloseModal={this.closeModal}
            />
          </div>
        ) :
        (
          <LoadingPage/>
        )}
      </div>
    );
  };
};


const mapStateToProps = (state, props) => ({
  expense: state.expenses.find((expense) => expense.id === props.match.params.id)
});

const mapDispatchToProps = (dispatch) => ({
  startEditExpense: (id, expense) => dispatch(startEditExpense(id, expense)),
  startRemoveExpense: (id) => dispatch(startRemoveExpense({ id }))
});

export default connect(mapStateToProps, mapDispatchToProps)(EditExpensePage);