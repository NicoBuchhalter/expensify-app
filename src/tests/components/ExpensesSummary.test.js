import React from 'react';
import { shallow } from 'enzyme';
import {ExpensesSummary} from '../../components/ExpensesSummary';

test('Should show correct message when only one expense', () => {
  expect(shallow(<ExpensesSummary allExpensesCount={1} expenseCount={1} expensesTotal={15000}/>)).toMatchSnapshot();
});

test('Should show correct message when multiple expenses', () => {
  expect(shallow(<ExpensesSummary allExpensesCount={5} expenseCount={3} expensesTotal={2550000}/>)).toMatchSnapshot();
});