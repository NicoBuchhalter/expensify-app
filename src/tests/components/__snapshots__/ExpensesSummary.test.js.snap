// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should show correct message when multiple expenses 1`] = `
<div
  className="page-header"
>
  <div
    className="content-container"
  >
    <h1
      className="page-header__title"
    >
      Viewing 
      <span>
        3
         of 
        5
      </span>
       
      expenses
       totalling 
      <span>
        $25,500.00
      </span>
    </h1>
    <div
      className="page-header__actions"
    >
      <Link
        className="button"
        to="/create"
      >
        Add Expense
      </Link>
    </div>
  </div>
</div>
`;

exports[`Should show correct message when only one expense 1`] = `
<div
  className="page-header"
>
  <div
    className="content-container"
  >
    <h1
      className="page-header__title"
    >
      Viewing 
      <span>
        1
         of 
        1
      </span>
       
      expense
       totalling 
      <span>
        $150.00
      </span>
    </h1>
    <div
      className="page-header__actions"
    >
      <Link
        className="button"
        to="/create"
      >
        Add Expense
      </Link>
    </div>
  </div>
</div>
`;
